staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -50,13 +50,17 @@ func IsListType(obj runtime.Object) bool {
 
 	isListCache.lock.RLock()
 	ok, exists := isListCache.byType[t]
+	lenByType := len(isListCache.byType)
 	isListCache.lock.RUnlock()
 
 	if !exists {
 		_, err := getItemsPtr(obj)
 		ok = err == nil
 
 		// cache only the first 1024 types
+		if lenByType >= 1024 {
```

Comment:
If this code change reduces lock contention, then the cache must be > 1024, no?

In which case something else must be wrong.
* I think we really should not have any where near that many types, since the runtime.Unstructured bit above should be catching all CRs.

If you tested this and the given change helps, then maybe we can consider deleting this cache entirely, converting it to an LRU, making it bigger, or figuring out why so many things are being put in it.

From the PR description it sounds like you're making this change based on just analyzing the code. I prefer to make changes like this only if there is a benchmark or some other demonstration that the change is an improvement--it can be counterintuitive what things make code run faster, and the benchmark prevents later regression.

Reply:
Good catch @lavalamp, I had just assumed that each CRD was a separate type here..

+1 to benchmarking and figuring out what is going on. And if we do need to limit cache, agree that LRU is much better than first 1024 types.

/lgtm cancel